src/control/security/signature.go

```diff
@@ -0,0 +1,104 @@
+//
+// (C) Copyright 2019 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+package security
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha512"
+	"io"
+
+	"github.com/pkg/errors"
+)
+
+//UnsupportedKeyError is a structured error used to indicate that the PublicKey
+//or PrivateKey interface passed in represents a key type we do not support.
+type UnsupportedKeyError struct{}
+
+//Error is the implementation of the error interface.
+func (err *UnsupportedKeyError) Error() string {
+	return "Key contains an unsupported key type"
+}
+
+//TokenSigner serves to encapsulate the functionality needed
+//to sign and verify auth token signatures.
+type TokenSigner struct {
+	randPool io.Reader
+}
+
+func DefaultTokenSigner() *TokenSigner {
+	return &TokenSigner{
+		randPool: rand.Reader,
+	}
+}
+
+//Sign takes an unhashed set of bytes and hashes and signs the result with the
+//key located in the file specified by PrivateKeyPath
+func (s *TokenSigner) Sign(key crypto.PrivateKey, data []byte) ([]byte, error) {
+
+	hash := sha512.New()
+	hash.Write(data)
```

Comment:
Write() returns an (int, error) tuple -- at a minimum the error value should be checked.

Reply:
I don't think it actually can. I think that's just a byproduct of the Writter interface. Not even the internal calls to Write in the the implementation of these hashes return their errors. The actual implementation of write just calls return at the end with no values. So you won't get anything back.

Reply:
Understood, but from a safety perspective you should still check the error. Relying on the implementation always returning a nil error isn't a good bet -- future versions of the stdlib might change things such that a non-nil error could be returned.

Reply:
Can you guarantee that err will always be nil? Its not clear what the semantics of a return line with no parameters are.

Reply:
Not sure what you're asking, exactly... If you're saying that the Write() implementation you're looking at ends with a return with no parameters, then the parameters are named in the function signature.

All I'm saying is that adding the following ensures that your code will never be in an undefined state if the implementation changes... It's a few extra lines of code for increased certainty:
```go
        if _, err := hash.Write(data); err != nil {
                return nil, errors.New("hash failed to write")
        }
```